File: API/src/main/java/com/ltizzi/ecommerce/model/cart/CartMapper.java
package com.ltizzi.ecommerce.model.cart;

import com.ltizzi.ecommerce.model.product.ProductEntity;
import com.ltizzi.ecommerce.model.product.ProductMapper;
import com.ltizzi.ecommerce.model.product.ProductResponse;
import com.ltizzi.ecommerce.model.user.UserEntity;
import com.ltizzi.ecommerce.model.user.UserMapper;
import com.ltizzi.ecommerce.model.user.UserResponse;
import com.ltizzi.ecommerce.repository.CartRepository;
import com.ltizzi.ecommerce.repository.ProductRepository;
import com.ltizzi.ecommerce.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import org.springframework.web.client.HttpClientErrorException;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

/**
 * @author Leonardo Terlizzi
 */
@Component
public class CartMapper {

//    @Autowired
//    private UserMapper userMapper;

    @Autowired
    private ProductMapper productMapper;

    @Autowired
    private ProductRepository prodRepo;

    @Autowired
    private CartRepository cartRepo;

    @Autowired
    private UserRepository userRepo;


    public CartResponse toCartResponse(CartEntity cart) {
        CartResponse cartRes = new CartResponse();
        cartRes.setId(cart.getCart_id());
        cartRes.setTotal(cart.getTotal());
        cartRes.setCantidad(cart.getCantidad());
        ProductResponse prod = productMapper.toProductResponse(cart.getProduct());
        cartRes.setProduct(prod);
        cartRes.setUser_id(cart.getUser().getUser_id());
        return cartRes;
    }

    public CartEntity toCartEntity(CartRequest cartReq) {
        CartEntity cart = new CartEntity();
        if (cartReq.getId() != null) {
            cart = cartRepo.findById(cartReq.getId()).orElseThrow();

        }
        cart.setUser(userRepo.findById(cartReq.getUser_id()).orElse(null));
        cart.setCantidad(cartReq.getCantidad());
        cart.setTotal(cartReq.getTotal());
        Long id = cartReq.getProduct().getId();
        Optional<ProductEntity> product = prodRepo.findById(id);
        cart.setProduct(product.get());
        return cart;
    }

    public CartEntity toCartEntity(CartResponse cartRes) {
        CartEntity cart = new CartEntity();
        if (cartRes.getId() != null) {
            cart = cartRepo.findById(cartRes.getId()).orElseThrow();
        }
        cart.setUser(userRepo.findById(cartRes.getUser_id()).orElse(null));
        cart.setCantidad(cartRes.getCantidad());
        cart.setTotal(cartRes.getTotal());
        Long id = cartRes.getProduct().getId();
        Optional<ProductEntity> product = prodRepo.findById(id);
        cart.setProduct(product.get());
        return cart;
    }

    public List<CartResponse> toArrayCartResponse(List<CartEntity> carts) {
        List<CartResponse> cartsRes = new ArrayList<>();
        carts.forEach(cart -> {
            cartsRes.add(toCartResponse(cart));
        });
        return cartsRes;
    }

    public List<CartEntity> toArrayCartEntity(List<CartRequest> cartsReq) {
        List<CartEntity> carts = new ArrayList<>();
        cartsReq.forEach(cart -> {
            carts.add(toCartEntity(cart));
        });
        return carts;
    }

    public List<CartEntity> toArrayCartEntityFromResponse(List<CartResponse> cartsRes) {
        List<CartEntity> carts = new ArrayList<>();
        cartsRes.forEach(cart -> {
            carts.add(toCartEntity(cart));
        });
        return carts;
    }
}

File: Server/src/dto/mappers/cart.mapper.ts
import { Cart } from "../../models/Cart";
import { User } from "../../models/User";
import { CartService } from "../../services/CartService";
import { UserService } from "../../services/UserService";
import { CartRequest } from "../requests/cart.request";
import { CartResponse } from "../responses/cart.response";
import { ProductMapper } from "./product.mapper";

const prodMapper = new ProductMapper();
const userServ = new UserService();
const cartServ = new CartService();

export class CartMapper {
  toCartResponse(cart: Cart): CartResponse {
    const cartRes = {} as CartResponse;
    // console.log("FROM MAPPER: ");
    // console.log(cart);
    cartRes.cart_id = cart.cart_id;
    cartRes.product = prodMapper.toProductResponse(cart.product);
    cartRes.cantidad = cart.cantidad;
    cartRes.total = cart.total;
    cartRes.user_id = cart.user_id;
    return cartRes;
  }

  async toCartEntity(fromCart: CartResponse | CartRequest): Promise<Cart> {
    let cart = {} as Cart;
    console.log("FROM MAPPER: ");
    console.log(fromCart);
    if (fromCart.cart_id) {
      cart = (await cartServ.getCartById(fromCart.cart_id)) as Cart;
    }
    cart.user = (await userServ.getUserById(fromCart.user_id)) as User;
    cart.user_id = fromCart.user_id;
    cart.cantidad = fromCart.cantidad;
    cart.total = fromCart.total;
    cart.product = await prodMapper.toProductEntity(fromCart.product);
    return cart;
  }

  toArrayCartResponse(carts: Array<Cart>): Array<CartResponse> {
    let cartsRes = [] as Array<CartResponse>;
    carts.forEach((cart) => cartsRes.push(this.toCartResponse(cart)));
    return cartsRes;
  }

  async toArrayCartEntity(
    fromCarts: Array<CartResponse | CartRequest>
  ): Promise<Array<Cart>> {
    let carts = [] as Array<Cart>;
    fromCarts.forEach(async (cart) =>
      carts.push(await this.toCartEntity(cart))
    );
    return carts;
  }
}
